### ISSUE TITLE
Second data connection fails

### ISSUE THREAD

**Issue by ws0w** [Castaglia/proftpd-mod_proxy#172]
I'm trying to use the latest mod_proxy with ProFTPd 1.3.7a doing a reverse proxy to a ProFTPd 1.3.6 backend server. Both servers are AIX 7.1.

After login in the the frontend ProFTPd, I can successfully execute one command (like dir or get) that uses a data connection. A second attempt hangs and them times out after 5 minutes. After the timeout, the following error message appears:
```
425 Unable to build data connection: Not owner
```
I've attached config, proftpd -V output, and a trace file produced the the following Trace options:
```
Trace proxy:30 \
    command:20 \
    proxy.conn:30 \
    proxy.db:30 \
    proxy.forward:30 \
    proxy.ftp.conn:30 \
    proxy.ftp.ctrl:30 \
    proxy.ftp.data:30 \
    proxy.ftp.msg:30 \
    proxy.ftp.sess:30 \
    proxy.ftp.xfer:30 \
    proxy.inet:30 \
    proxy.netio:30 \
    proxy.random:30 \
    proxy.reverse:30 \
    proxy.reverse.db:30 \
    proxy.reverse.redis:30 \
    proxy.session:30 \
    proxy.tls:30 \
    proxy.tls.db:30 \
    proxy.tls.redis:30 \
    proxy.uri:30 \
    response:20
```
Let me know what other information would be helpful.

[issue172.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5234591/issue172.tar.gz)

**Comment by Castaglia** [comment]
When you attempt the second data transfer (a directory listing, for the provided logs), you see this response in the client, yes?
```
2020-09-16 13:38:55,570 [10289224] <proxy:19>: read EOF on data connection, closing frontend/backend data connections
2020-09-16 13:38:55,570 [10289224] <proxy.netio:18>: (inet_close) found tls data NetIO
2020-09-16 13:38:55,570 [10289224] <proxy.netio:18>: (netio_read) found tls ctrl NetIO
2020-09-16 13:38:55,570 [10289224] <proxy.netio:19>: (netio_read) using proxy ctrl NetIO
2020-09-16 13:38:55,570 [10289224] <proxy.ftp.ctrl:9>: received '425 Unable to build data connection: Not owner' response from backend to frontend
2020-09-16 13:38:55,570 [10289224] <proxy.ftp.ctrl:9>: backend->frontend response: 425 Unable to build data connection: Not owner
2020-09-16 13:38:55,571 [10289224] <response:1>: 425 Unable to build data connection: Not owner
```
I've not seen this "Not owner" error before; will see what I can find out.

Sorry for the delay in looking into this.

**Comment by Castaglia** [comment]
Do you happen to have the logs from the backend ProFTPD server, so we can see why it might be rejecting the second data connection?

**Comment by ws0w** [comment]
Sorry, I just saw your response. I'll have to recreate to get a log from the backend.

And, yes, the log lines you pasted above are from the client when attempting the second data connection, for the 'dir' command.

**Comment by ws0w** [comment]
The attached archive contains two trace files. One from the frontend and one from the backend.

I logged in to the frontend and issued the 'dir' command twice, then waited for the second 'dir' command to timeout. Sorry again for the long delay.
 
[trace.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5718221/trace.tar.gz)

**Comment by ws0w** [comment]
This line from the frontend log is interesting:

2020-12-18 14:34:03,171 [3211446] <proxy.ftp.conn:9>: connecting to 9.16.18.45#65189 from 9.162.144.202#**21**

The connection from 9.162.144.202 to 9.16.18.45#65189 was not from port 21.

I've been testing this over and over, and at that point in the communication, the log always incorrectly mentions port 21 as the source port instead of the actually source port.

**Comment by Castaglia** [comment]
The port number in the "from" address here:
```
proxy.ftp.conn:9: connecting to 9.16.18.45#65189 from 9.162.144.202#21
```
is indeed a bit misleading.  Currently, what is being logged is the IP address and port of the _listening address_ contacted by the frontend client.  I'll see if I can improve the accuracy of that source port in the log message.  Still, that shouldn't be causing any actual connection issues, per se -- that is, I'm hoping that this particular log message inaccuracy is just in the logging.

**Comment by Castaglia** [comment]
Just some notes for my future reference on this:
* frontend client requests _active_ data transfers (via `PORT` command) from proxy
* proxy requests _passive_ data transfers (via `EPSV` command) from backend server
* TCP connect from proxy to frontend client for data transfers succeed
* TCP connect from proxy to backend server for data transfers succeed
* Second/subsequent data IO (TLS handshakes, data transfers) from backend server to proxy _fail_ with "425 Unable to build data connection: Not owner"

@ws0w I presume that the backend server (another ProFTPD server, it looks like) works successfully, for multiple data transfers (_e.g._ directory listings) when used directly (not via proxy)?  That is, if you use an FTP client to request the same directory listings (using `EPSV`, `LIST` commands) from the backend server directly, they work?

If using an FTP client directly against the backend server works, it'd be useful to get _that_ trace logging, to compare against what `mod_proxy` is doing, to see what might be different.

**Comment by Castaglia** [comment]
Another fun little tidbit of learning for my future self: _some_ platforms log the `EPERM` errno value as "Operation not permitted", and _others_ log it as "Not owner".  Thus the issues here may not actually be one of _ownership_, per se, but something else.  I've become accustomed to Linux, which uses `EACCES` ("Access denied") for ownership/access errors.

**Comment by Castaglia** [comment]
@ws0w In addition to the trace logging, would it be possible to get the system and debug logging for the backend server for these failed transfers?  That is:
```
SystemLog /path/to/proftpd-trace.log
DebugLevel 10
```
as well?  I think there may be some additional clues in that logging, that is not in the trace logging.  Thanks!

**Comment by ws0w** [comment]
Yes, multiple data connections work just fine going directly from client to backend (also ProFTPd on AIX, same binary as frontend). I even tried multiple data connections and switching back and forth between acitive and passive.

When going through the proxy, seems to fail the same way regardless of active/passive combinations.

I'll work on producing the requested logs.

**Comment by ws0w** [comment]
I've attached the logs ... the backend's systemLog (backend.proftpd-trace.log) contains only the following three lines (which correspond to direct activity, without going throught the proxy):

>2020-12-21 13:00:34,907 dhemddel01 proftpd[19988488] dispmb-209.mul.ie.ibm.com (wilson.boulder.ibm.com[9.16.18.45]): error writing format string '[msg] %.*s' into 4048-byte buffer: No space left on device
>2020-12-21 13:00:43,237 dhemddel01 proftpd[19988488] dispmb-209.mul.ie.ibm.com (wilson.boulder.ibm.com[9.16.18.45]): error writing format string '[msg] %.*s' into 4048-byte buffer: No space left on device
>2020-12-21 13:00:45,245 dhemddel01 proftpd[19988488] dispmb-209.mul.ie.ibm.com (wilson.boulder.ibm.com[9.16.18.45]): error writing format string '[msg] %.*s' into 4048-byte buffer: No space left on device


There are no filesystem issues on the server.

[backendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5726397/backendlogs.tar.gz)
[frontendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5726398/frontendlogs.tar.gz)

**Comment by ws0w** [comment]
It seems that DebugLevel doen't do much unless it's in a Global or VirtualHost stanza.

**Comment by Castaglia** [comment]
Based on the timestamps, these log messages:
```
2020-12-21 13:00:45,245 dhemddel01 proftpd[19988488] dispmb-209.mul.ie.ibm.com (wilson.boulder.ibm.com[9.16.18.45]): error writing format string '[msg] %.*s' into 4048-byte buffer: No space left on device
```
Fit into these log messages:
```
2020-12-21 13:00:45,245 mod_tls/2.9[19988488]: [info] accepting: SSLv3 write server done A
2020-12-21 13:00:45,245 mod_tls/2.9[19988488]: [info] accepting: SSLv3 flush data
2020-12-21 13:00:45,245 mod_tls/2.9[19988488]: [info] accepting: SSLv3 read client certificate A
2020-12-21 13:00:45,245 mod_tls/2.9[19988488]: [info] accepting: SSLv3 read client key exchange A
2020-12-21 13:00:45,245 mod_tls/2.9[19988488]: [info] accepting: SSLv3 read client key exchange A
```
which makes sense.  The above shows that the TLS session between proxy->backend server uses client auth.  The logged message is trying to show the data about that client cert -- which apparently exceeds the 4KB log message buffer size, hence the `error writing format string '[msg] %.*s' into 4048-byte buffer`.  Fortunately, this is not the cause of the problem.

_That_ happens later:
```
2020-12-21 13:01:36,651 [5963878] <command:7>: dispatching CMD command 'LIST' to mod_ls.c
2020-12-21 13:01:36,652 [5963878] <tls:5>: ignoring SNI for data connections
2020-12-21 13:01:36,652 [5963878] <tls:17>: calling SSL_accept() on data conn fd 15
...
2020-12-21 13:06:36,700 mod_tls/2.9[5963878]: TLS negotiation timed out (300 seconds)
2020-12-21 13:06:36,701 mod_tls/2.9[5963878]: unable to open data connection: TLS negotiation failed
```
This is the backend server, wanting to accept the TLS handshake (from the proxy) for the second data transfer -- and timing out.

@ws0w Am I correct in assuming that this data transfer failure happens when the frontent client is using FTPS, for both control _and_ data connections, to the proxy?  Does it happen when the frontend client is _not_ using TLS for the data transfers to the proxy, only for the control connection?  I know it sounds like a weird question, but `mod_proxy` does some fancy juggling, in order to work with/around `mod_tls`, for handling of separate TLS sessions on the frontend (to the client) _and_ on the backend (to the server).  I'm wondering if there is something amiss, in that juggling, which your setup is tickling...so I'm trying to re-create the _exact_ setup locally.

**Comment by ws0w** [comment]
I haven’t tried not using TLS for the data connection. I’ll try that in the morning.

**Comment by ws0w** [comment]
The second data connection also hangs when NOT using TLS for the data connection between the frontend client and the proxy.

**Comment by Castaglia** [comment]
@ws0w OK, that's very good to know.  Any chance of getting detailed logs, trace logs, from both the proxy and the backend server, for the non-TLS data transfer cases?

**Comment by ws0w** [comment]
Here they are:

[backendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5731989/backendlogs.tar.gz)
[frontendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5731990/frontendlogs.tar.gz)

**Comment by Castaglia** [comment]
Is it possible get these traces where there is no TLS between the proxy and the backend server?  I suspect that this issue is between `mod_proxy`, and the backend server.  I'm not sure if TLS, on these proxy/backend data transfers, is a factor or not.

**Comment by ws0w** [comment]
Sorry about the extra tracing ... I thought I removed those before running that test.
I only added some trace logging.

I'm working getting logs for a clear data connection between proxy and backend ... I can't even get one data connection with that scenario.

**Comment by ws0w** [comment]
Here are the logs with proxy and backend configured with:

>TLSRequire auth

and the proxy configured with

>ProxyTLSTransferProtectionPolicy clear

The logs contain two ftp sessions. In the first session, I issued the 'clear' command from the client before issueing a 'dir' command. The dir command hung. 
In the second session, I let the client to default to an encrypted data connection. The 'dir' command hung for that session also.

[backendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5736637/backendlogs.tar.gz)
[frontendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5736638/frontendlogs.tar.gz)

**Comment by ws0w** [comment]
Here are logs of two more ftp sessions. The difference from the previous session in that on the proxy server, I used this setting:

>ProxyTLSTransferProtectionPolicy client

With that proxy setting, when I issue 'clear' from the client, the 'dir' command is successful, and works multiple times.
When I let the client default to an encrypted data session, the 'dir' command fails on the first try. But, it doesn't hang ... I get this error message immediately:

>ERROR  during TLS handshake for data connection

**Comment by Castaglia** [comment]
Thanks for those; having logs for cases that work all the time definitely help, for providing comparisons.

Are you, by change, using version 0.6 of `mod_proxy`, or 0.7?  If you are using version 0.6, would you be able to try the 0.7 version?

**Comment by ws0w** [comment]
It looks like I have version 0.6.  I cloned the current repository and that got me version 0.8 ... is that what you want me to try? I don't see version 0.7.

**Comment by Castaglia** [comment]
Yeah, the latest `mod_proxy` from GitHub is best to use, thanks.

**Comment by ws0w** [comment]
I recompiled with the latest mod_proxy, but the behavior is unchanged.

**Comment by Castaglia** [comment]
@ws0w Hmm, OK, thanks.  Would it be possible to get the detail logs (trace, debug, module-specific) using the latest `mod_proxy` code?  There have been some logging changes, and I'm hoping to get a few more clues...

**Comment by ws0w** [comment]
Ok, I'll get some more logs.

**Comment by ws0w** [comment]
Here are logs for the following scenario:

Client defaults to encrypted data

Frontend proxy

>TLSRequire auth
>ProxyTLSTransferProtectionPolicy default (required)

Backend

>TLSRequire auth

[backendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5752402/backendlogs.tar.gz)
[frontendlogs.tar.gz](https://github.com/Castaglia/proftpd-mod_proxy/files/5752403/frontendlogs.tar.gz)

**Comment by Castaglia** [comment]
OK, I think I have a hypothesis which explains the behaviors you've seen.

First, the initial "TLS handshake timed out", for the second data transfer:
```
2020-12-29 15:15:57,087 mod_tls/2.9[17432586]: TLS negotiation timed out (300 seconds)
2020-12-29 15:15:57,087 mod_tls/2.9[17432586]: unable to open data connection: TLS negotiation failed
```
We saw that `mod_proxy` made the proper TCP connections for these active data transfers (`PORT`, `LIST`) to the frontend and backend hosts.  But for the backend host, it timed out waiting for `mod_proxy` to perform the TLS handshake for this second transfer.  And `mod_proxy` didn't.  Instead, it was _not_ actually using the TLS-related NetIO callbacks for this subsequent transfer; it was just using the normal TCP NetIO callbacks.  Thus `mod_proxy` was waiting for the backend to send the requested `LIST` data.  Both sides waiting for the other to send data first -- deadlock, and timeout.  The 300 seconds is the default [`TLSTimeoutHandshake` (http://www.proftpd.org/docs/contrib/mod_tls.html#TLSTimeoutHandshake) value.

This happens because of the code added https://github.com/Castaglia/proftpd-mod_proxy/commit/4607dfb556a5f6679825aa4d52d4096d25ecd258#diff-e1ebe896ec4a6aa05982a001168a8dfff0bcb70a15d75f9368b771ff94fc559d -- specifically, when `mod_proxy` closed the _first_ TLS data transfer connection, the above code "popped the NetIO stack" improperly, removing the TLS-related NetIO callbacks.  And thus the initial TLS data transfer succeeded, but all subsequent ones fail.

Interestingly enough, I encountered (and fixed) the above behavior while working on https://github.com/Castaglia/proftpd-mod_proxy/issues/190 -- only there, for me, this improper NetIO stack popping caused a segfault (due to null pointer dereference) for me locally.

Next, the third/fourth data transfers fail _immediately, not even waiting for 300 seconds, because of a `mod_tls` bug, now filed as https://github.com/proftpd/proftpd/issues/1204.

So, for the current state of things, I'd recommend trying out the latest `mod_proxy` code (and latest ProFTPD source from the `master` branch there), if you can.  Hopefully these subsequent proxied data transfers, using TLS, will work better for you now.

**Comment by ws0w** [comment]
Thank, I’ll try it out.

**Comment by pedanticdan** [comment]
I finally got. chance to work on this, but I can't figure out how to build the latest proftpd from master.

I get this message from ./configure:

configure: WARNING: unrecognized options: --enable-openssl, --enable-ctrls, --disable-auth-pam, --with-modules, --disable-shared, --disable-dso, --enable-sendfile, --disable-strip, --enable-tunable-buffer-size

I'm not sure if something changed in the build process since 1.3.7a, or if I made some stupid mistake.
